File: src/Overseer/Machines/Providers/OctoprintMachineProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

using Overseer.Models;

namespace Overseer.Machines.Providers
{
  public class OctoprintMachineProvider : PollingMachineProvider<OctoprintMachine>
  {
    public OctoprintMachineProvider(OctoprintMachine machine)
    {
      Machine = machine;
    }

    protected override Task<dynamic> Fetch(string resource, FetchRequest request = default, CancellationToken cancellation = default)
    {
      request ??= new();
      request.Headers.TryAdd("X-Api-Key", Machine.ApiKey);
      return base.Fetch(resource, request, cancellation);
    }

    /// <summary>
    /// This uses the Octoprint API to pause the print,
    /// this is needed because the Gcode command only works
    /// when printing from SD, where as when printing from 
    /// Octoprint the commands are streamed to the printer.
    /// </summary>
    public override Task PauseJob()
    {
      return Fetch("api/job", new() { Method = "POST", Body = new { command = "pause", action = "pause" } });
    }

    /// <summary>
    /// This uses the Octoprint API to resume the print,
    /// this is needed because the Gcode command only works
    /// when printing from SD, where as when printing from 
    /// Octoprint the commands are streamed to the printer.
    /// </summary>
    public override Task ResumeJob()
    {
      return Fetch("api/job", new() { Method = "POST", Body = new { command = "pause", action = "resume" } });
    }

    /// <summary>
    /// This uses the Octoprint API to cancel the print,
    /// this is needed because the Gcode command only works
    /// when printing from SD, where as when printing from 
    /// Octoprint the commands are streamed to the printer.
    /// </summary>
    public override Task CancelJob()
    {
      return Fetch("api/job", new() { Method = "POST", Body = new { command = "cancel" } });
    }

    public override Task ExecuteGcode(string command)
    {
      return Fetch("api/printer/command", new() { Method = "POST", Body = new { command } });
    }

    public override async Task LoadConfiguration(Machine machine)
    {
      try
      {
        OctoprintMachine updatedMachine = machine as OctoprintMachine;
        //the api path will be auto generated, so make sure the machine has the root path
        updatedMachine.Url = ProcessUri(updatedMachine.Url, string.Empty).ToString();

        var settings = await Fetch("api/settings");
        if (string.IsNullOrWhiteSpace(updatedMachine.WebCamUrl))
        {
          updatedMachine.WebCamUrl = ProcessUri(updatedMachine.Url, (string)settings["webcam"]["streamUrl"]).ToString();

          if ((bool)settings["webcam"]["flipH"])
          {
            updatedMachine.WebCamOrientation = WebCamOrientation.FlippedHorizontally;
          }
          else if ((bool)settings["webcam"]["flipV"])
          {
            updatedMachine.WebCamOrientation = WebCamOrientation.FlippedVertically;
          }
        }

        updatedMachine.AvailableProfiles.Clear();
        dynamic profiles = await Fetch("api/printerprofiles");
        foreach (dynamic profileProperty in profiles["profiles"])
        {
          var profile = profileProperty.Value;
          updatedMachine.AvailableProfiles.Add((string)profile.id, (string)profile.name);

          if ((bool)profile.current)
          {
            var tools = new List<MachineTool>();
            updatedMachine.Profile = profile.name;

            if ((bool)profile.heatedBed)
            {
              tools.Add(new MachineTool(MachineToolType.Heater, -1, "bed"));
            }

            int extruderCount = profile.extruder.count;
            bool sharedNozzle = profile.extruder.sharedNozzle;

            if (sharedNozzle)
            {
              tools.Add(new MachineTool(MachineToolType.Heater, 0));
            }

            for (int i = 0; i < extruderCount; i++)
            {
              if (!sharedNozzle)
              {
                tools.Add(new MachineTool(MachineToolType.Heater, i));
              }

              tools.Add(new MachineTool(MachineToolType.Extruder, i));
            }

            updatedMachine.Tools = tools;
          }
        }

        Machine = updatedMachine;
      }
      catch (Exception ex)
      {
        Log.Error("Load Configuration Failure", ex);

        //checks if the exception, or any inner exceptions, is an overseer exception and if so throws it
        OverseerException.Unwrap(ex);

        if (ex.Message.Contains("Invalid API key"))
          throw new OverseerException("octoprint_invalid_key", Machine);

        throw new OverseerException("machine_connect_failure", Machine);
      }
    }

    private int GetHeaterIndex(string heaterKey)
    {
      if (heaterKey.ToLower() == "bed") return -1;
      return Convert.ToInt32(heaterKey.Replace("tool", string.Empty));
    }

    protected override async Task<MachineStatus> AcquireStatus(CancellationToken cancellationToken)
    {
      dynamic machineState = await Fetch("api/printer", cancellation: cancellationToken);
      var status = new MachineStatus { MachineId = MachineId };

      foreach (dynamic pair in machineState.temperature)
      {
        var heaterIndex = GetHeaterIndex(pair.Name);
        status.Temperatures.Add(heaterIndex, new TemperatureStatus
        {
          HeaterIndex = heaterIndex,
          Actual = pair.Value.actual,
          Target = pair.Value.target
        });
      }

      var flags = machineState.state.flags;
      if ((bool)flags.operational)
      {
        if ((bool)flags.paused || (bool)flags.pausing)
        {
          status.State = MachineState.Paused;
        }
        else if ((bool)flags.printing || (bool)flags.resuming)
        {
          status.State = MachineState.Operational;
        }
        else
        {
          status.State = MachineState.Idle;
        }

        var jobStatus = await Fetch("api/job", cancellation: cancellationToken);
        status.ElapsedJobTime = (int?)jobStatus["progress"]["printTime"] ?? 0;
        status.EstimatedTimeRemaining = (int?)jobStatus["progress"]["printTimeLeft"] ?? 0;
        status.Progress = (float)Math.Round((float?)jobStatus["progress"]["completion"] ?? 0f, 1);

        // these values can't be retrieved through the octoprint api, or from marlin by gcode, so it's not going
        // to be supported by the Octoprint provider, however the values are being defaulted here to replicate the 
        // Octoprint UI behavior upon refresh. 
        status.FanSpeed = 100;
        status.FeedRate = 100;
        status.FlowRates = Machine.Tools
            .Where(tool => tool.ToolType == MachineToolType.Extruder)
            .ToDictionary(tool => tool.Index, tool => 100f);
      }
      else
      {
        status.State = MachineState.Idle;
      }

      return status;
    }
  }
}

File: src/Overseer.Client/src/app/services/local/providers/octoprint-machine.provider.ts
import { HttpClient } from '@angular/common/http';
import { forkJoin, Observable, of, throwError } from 'rxjs';
import { catchError, map, mergeMap } from 'rxjs/operators';

import { MachineStatus } from '../../../models/machine-status.model';
import { Machine, MachineTool, OctoprintMachine } from '../../../models/machine.model';
import { PollingMachineProvider } from './machine.provider';
import { processUrl } from './url-processor';

export class OctoprintMachineProvider extends PollingMachineProvider<OctoprintMachine> {
  constructor(
    machine: Machine,
    private http: HttpClient
  ) {
    super();
    this.machine = machine as OctoprintMachine;
  }

  get httpOptions(): { headers?: { [header: string]: string | string[] } } {
    return { headers: { 'X-Api-Key': this.machine.apiKey ?? '' } };
  }

  override pauseJob(): Observable<void> {
    return this.http.post<void>(processUrl(this.machine.url, 'api/job'), { command: 'pause', action: 'pause' }, this.httpOptions);
  }

  override resumeJob(): Observable<void> {
    return this.http.post<void>(processUrl(this.machine.url, 'api/job'), { command: 'pause', action: 'resume' }, this.httpOptions);
  }

  override cancelJob(): Observable<void> {
    return this.http.post<void>(processUrl(this.machine.url, 'api/job'), { command: 'cancel' }, this.httpOptions);
  }

  override executeGcode(command: string): Observable<void> {
    return this.http.post<void>(processUrl(this.machine.url, 'api/printer/command'), { command: command }, this.httpOptions);
  }

  loadConfiguration(machine: Machine): Observable<Machine> {
    if (machine.machineType !== 'Octoprint') return of(machine);

    machine.url = processUrl(machine.url);

    return forkJoin([
      this.http.get<any>(processUrl(machine.url, 'api/settings'), this.httpOptions),
      this.http.get<any>(processUrl(machine.url, 'api/printerprofiles'), this.httpOptions),
    ]).pipe(
      map(([settings, profiles]) => {
        if (!machine.webCamUrl) {
          machine.webCamUrl = processUrl(machine.url, settings.webcam.streamUrl);

          if (settings.webcam.flipH) {
            machine.webCamOrientation = 'FlippedHorizontally';
          } else if (settings.webcam.flipV) {
            machine.webCamOrientation = 'FlippedVertically';
          } else {
            machine.webCamOrientation = 'Default';
          }
        }

        machine.availableProfiles = new Map();
        for (const profileKey in profiles.profiles) {
          if (!profiles.profiles.hasOwnProperty(profileKey)) {
            continue;
          }

          const profile = profiles.profiles[profileKey];
          machine.availableProfiles.set(profile.id, profile.name);

          if (profile.current) {
            const tools: MachineTool[] = [];

            if (profile.heatedBed) {
              tools.push({
                toolType: 'Heater',
                index: -1,
                name: 'bed',
              });
            }

            if (profile.extruder.sharedNozzle) {
              tools.push({
                toolType: 'Heater',
                index: 0,
                name: 'heater 0',
              });
            }

            for (let index = 0; index < profile.extruder.count; index++) {
              if (!profile.extruder.sharedNozzle) {
                tools.push({
                  toolType: 'Heater',
                  index: index,
                  name: `heater ${index}`,
                });
              }

              tools.push({
                toolType: 'Extruder',
                index: index,
                name: `extruder ${index}`,
              });
            }

            machine.tools = tools;
            machine.profile = profile.name;
          }
        }

        this.machine = machine;
        return machine;
      }),
      catchError((err: Error) => {
        if (err.message.indexOf('Invalid API key') >= 0) {
          return throwError(() => new Error('octoprint_invalid_key'));
        }

        return throwError(() => new Error('machine_connect_failure'));
      })
    );
  }

  getHeaterIndex(heaterKey: string): number {
    if (heaterKey.toLowerCase() === 'bed') {
      return -1;
    }
    return parseInt(heaterKey.replace('tool', ''), 10);
  }

  acquireStatus(): Observable<MachineStatus> {
    const status: MachineStatus = {
      machineId: this.machine.id,
      state: 'Idle',
      temperatures: {},
    };

    return this.http.get<any>(processUrl(this.machine.url, 'api/printer'), this.httpOptions).pipe(
      mergeMap((machineState) => {
        status.temperatures = {};
        for (const key in machineState.temperature) {
          const temp = machineState.temperature[key];
          const index = this.getHeaterIndex(key);

          status.temperatures[index] = {
            heaterIndex: index,
            actual: temp.actual,
            target: temp.target,
          };
        }

        const flags = machineState.state.flags;
        if (flags.paused || flags.pausing) {
          status.state = 'Paused';
        } else if (flags.printing || flags.resuming) {
          status.state = 'Operational';
        } else {
          status.state = 'Idle';
        }

        if (status.state === 'Operational' || status.state === 'Paused') {
          return this.http.get<any>(processUrl(this.machine.url, 'api/job'), this.httpOptions).pipe(
            map((jobStatus) => {
              status.elapsedJobTime = jobStatus.progress.printTime;
              status.estimatedTimeRemaining = jobStatus.progress.printTimeLeft;
              status.progress = jobStatus.progress.completion;
              status.fanSpeed = 100;
              status.feedRate = 100;
              status.flowRates = this.machine.tools
                .filter((tool) => tool.toolType === 'Extruder')
                .reduce<{ [key: number]: number }>((obj, tool) => {
                  obj[tool.index] = 100;
                  return obj;
                }, {});

              return status;
            })
          );
        }

        return of(status);
      })
    );
  }
}
